File: LISST_VSF_correction.py






import numpy as np
from matplotlib import pyplot as plt
import math
import scipy.integrate as integrate

pi=math.pi

nrOfInputs=158
layers=4

#######   Path to directory containing model

Filepath = 'Insert path'

#######   import transformation values   ########

avg_in=np.loadtxt(Filepath + "/avg_in.txt")
avg_out=np.loadtxt(Filepath + "/avg_out.txt")
std_in=np.loadtxt(Filepath + "/std_in.txt")
std_out=np.loadtxt(Filepath + "/std_out.txt")

###### Import angles #######

angles=np.loadtxt(Filepath + "/Angles.txt")

#####  Path and file containg input VSF     #########
    
VSF_matrix=np.loadtxt(Filepath + "VSF_example.txt")
input_matrix=VSF_matrix[0:158,:]   ##### Only want input VSF up to 140 degrees


nrOfSamples=len(input_matrix[0])    


######## Apply transformation   ##########

input_matrix=np.log(input_matrix)
input_matrix=input_matrix.transpose()
input_matrix=(input_matrix-avg_in)/std_in

#####  Import weights and bias   ###########

W=[]
B=[]

for i in range (layers):
    W.append(np.loadtxt(Filepath + 'Weights_'+str(i+1)+'.txt'))
    B.append(np.loadtxt(Filepath + 'Bias_'+str(i+1)+'.txt'))

########   Neural network correction funtion   ##########

def activation(x):

    data = [math.tanh(value) for value in x] #0.5*(np.exp(value)-1) 

    return np.array(data, dtype=float)


def predict(input_matrix, layers, W, B):
        layer=input_matrix
        for i in range(layers): 
            if i==layers-1:
#                    print(len(layer))
                layer=np.dot(layer, W[i]) + B[i]
#                    print(np.dot(layer, W[i]) + B[i])
            else:
                layer = activation(np.dot(layer, W[i]) + B[i])
#                print(len(layer))
            
        prediction = layer
        
        return prediction        

        
###### Make correction  #####
        
prediction=[]    
    
for sample in range(nrOfSamples):        
    prediction.append(predict(input_matrix[sample], layers, W, B))

    
######### Apply reverse transformation    #############    
    
prediction=np.array(prediction) 
prediction=prediction.reshape(nrOfSamples, 168)

prediction=prediction*std_out+avg_out
prediction=np.exp(prediction)
prediction=prediction.transpose()

############# Make Figure ##############

fig, ax = plt.subplots()
ax.plot(angles, VSF_matrix[0:168,:], 'r', label='A')
ax.plot(angles, prediction[:,:], 'b', label='B')

y_label=ax.set_ylabel('VSF ['r'$m^{-1}sr^{-1}$]')
x_label=ax.set_xlabel(r'$\theta$ ''[degrees]')

handles, labels = ax.get_legend_handles_labels()
display = (0,nrOfSamples)
ax.legend([handle for i,handle in enumerate(handles) if i in display],['LISST-VSF', 'ANN'], loc=2)
plt.yscale('log')

x_label.set_fontsize(22)
y_label.set_fontsize(22)
ax.set_ylim(10**(-5), 10**6)
ax.set_frame_on(True)

plt.show()

################ Calculated scattering and backscattering coefficients #############   
 
ang_rad=angles*(pi/180)
ang_bb=ang_rad[107::]

b_all=np.zeros([nrOfSamples,2])
bb_all=np.zeros([nrOfSamples,2])
VSF_all_err=np.zeros([nrOfSamples,2])

for sample in range(0,nrOfSamples):

    b_all[sample,0]=(integrate.trapz(np.multiply(VSF_matrix[0:168,sample],2*pi*np.sin(ang_rad)), ang_rad))
    b_all[sample,1]=(integrate.trapz(np.multiply(prediction[:,sample],2*pi*np.sin(ang_rad)), ang_rad))

    bb_all[sample,0]=(integrate.trapz(np.multiply(VSF_matrix[107:168,sample],2*pi*np.sin(ang_bb)), ang_bb))
    bb_all[sample,1]=(integrate.trapz(np.multiply(prediction[107:,sample],2*pi*np.sin(ang_bb)), ang_bb))

##############  Prints scattering and backscattering coefficients  #############

np.set_printoptions(suppress=True, precision=2)    #### Sets number of displayed decimals
print(b_all)
np.set_printoptions(suppress=True, precision=4)
print(str(bb_all))

File: LISST_VSF_correction.m

%%%% Path to folder

Path='\\hallingskeid.uib.no\hug043\Documents\MonteCarlo\LISST_VSF\LISST_correction\Neural\NeuralNetwork\';

%%%%%%%% VSF to be corrected

VSF = readmatrix('\\hallingskeid.uib.no\hug043\Documents\MonteCarlo\LISST_VSF\LISST_correction\Neural\Koestner\DilCor\Lagoon_pro.txt');
numberOfSamples=length(VSF(1,:));
%%%%LISST-VSF angles

Angles=readmatrix(sprintf('%sAngles.txt',Path));

%%%% Load weights and biases

%%%% Weights

W{1}=readmatrix(sprintf('%sWeights_1.txt',Path));
W{2}=readmatrix(sprintf('%sWeights_2.txt',Path));
W{3}=readmatrix(sprintf('%sWeights_3.txt',Path));
W{4}=readmatrix(sprintf('%sWeights_4.txt',Path));

%%%%% Bias

B{1}=readmatrix(sprintf('%sBias_1.txt',Path));
B{2}=readmatrix(sprintf('%sBias_2.txt',Path));
B{3}=readmatrix(sprintf('%sBias_3.txt',Path));
B{4}=readmatrix(sprintf('%sBias_4.txt',Path));



%%%% Transformation values

avg_in=readmatrix(sprintf('%savg_in.txt',Path));
avg_out=readmatrix(sprintf('%savg_out.txt',Path));
std_in=readmatrix(sprintf('%sstd_in.txt',Path));
std_out=readmatrix(sprintf('%sstd_out.txt',Path));

%%%%% Transform input

VSF_input=VSF(1:158,:);    %%%%% Only want input for angles up to 140 degree

VSF_input=log(VSF_input);
VSF_input=(VSF_input-avg_in)./std_in;



%%%%% Make correction
layer=VSF_input;
for i=1:4
    Weights=W{i};
    Bias=B{i};
    new_layer=zeros(length(Bias), length(layer(1,:)));

    if i==4
        for j=1:length(layer(1,:))
            for k=1:length(Bias)
                new_layer(k,j)=dot(layer(:,j),Weights(:,k))+Bias(k);
            end
        end
    else
        for j=1:length(layer(1,:))
            for k=1:length(Bias)
                new_layer(k,j)=tanh(dot(layer(:,j),Weights(:,k))+Bias(k));
            end
        end    
    end
    layer=new_layer;

end

Prediction=layer;


%%%%%  Transform prediction

Prediction = Prediction .* std_out + avg_out;
Prediction = exp(Prediction);

%%%% Calculate scattering and backscattering coefficient

b_LISST=[];
bb_LISST=[];
b_NN=[];
bb_NN=[];
Angles_rad=Angles*pi/180;

for i=1:numberOfSamples
    
    %%%% LISST
    
    b_LISST = [b_LISST 2*pi*trapz(Angles_rad(1:168), VSF(1:168,i).*sin(Angles_rad(1:168)))];
    bb_LISST = [bb_LISST 2*pi*trapz(Angles_rad(108:168), VSF(108:168,i).*sin(Angles_rad(108:168)))];
    
    %%%% Neural network
    
    b_NN = [b_NN 2*pi*trapz(Angles_rad(1:168), Prediction(1:168,i).*sin(Angles_rad(1:168)))];
    bb_NN = [bb_NN 2*pi*trapz(Angles_rad(108:168), Prediction(108:168,i).*sin(Angles_rad(108:168)))];
end

%%%% Plot

figure();
    hold on;

    Pred=plot(Angles(1:length(Prediction)), Prediction,'Color',[0.7 0 0],'LineWidth', 1);
    LISST=plot(Angles(1:length(VSF)), VSF,'Color',[0 0 0],'LineWidth', 1);

    set(gca, 'YScale', 'log')
    xlabel('\theta (degrees)') 
    ylabel('VSF (m^{-1} sr^{-1})')
    legend([Pred(1), LISST(1)],{'Prediction', 'LISST-VSF'})

    box on;
